### PR TITLE
Adds support for loading NuGet dependencies for C# Projects

### PR DESCRIPTION
**Pull request by joncloud**
I've created a partial request to start supporting the ability to load NuGet dependencies for C# Projects. I wanted to get what I had out and visible to allow community review.

- [x] Implements the ability to search GitHub repositories using the filename pattern matching
  - [X] Anonymous searching
  - [x] Authenticated searching
- [x] Adds functions for parsing NuGet dependencies for C# projects
  - [X] *.csproj format
  - [x] packages.config format
- [ ] ~~Accepts uploads for parsing NuGet dependencies~~

### CODE REVIEW

**Comment by joncloud** [comment]
Here are a couple of GitHub accounts that have csproj formatted NuGet dependencies along with a few specific examples (both accounts have quite a bit more):
* [joncloud](https://github.com/joncloud)
  * [egad-net](https://github.com/joncloud/egad-net)
  * [apogee-net](https://github.com/joncloud/apogee-net)
  * [amp-net](https://github.com/joncloud/amp-net)
* [aspnet](https://github.com/aspnet)
  * [EntityFrameworkCore](https://github.com/aspnet/EntityFrameworkCore)
  * [Blazor](https://github.com/aspnet/Blazor)
  * [Mvc](https://github.com/aspnet/Mvc)

I know that a recent platform just started called [Discover .NET](https://discoverdot.net/), which is a showcase for open source .NET projects (C# is a subset).

File: src/lib/github.js [hunk 1]
```diff
@@ -119,6 +120,27 @@ async function fetchReposForProfile (profile, accessToken) {
   return accessToken ? repos : publicRepos;
 }
 
+function searchFilesFromRepo (repo, searchPattern, accessToken) {
```
Comment (joncloud) [inline src/lib/github.js]: I'm not sure if we can leverage `octokit` to execute this query as opposed to re-implementing it here.

**Comment by joncloud** [comment]
After some more research I think I would be more inclined to introduce uploading NuGet dependencies through a separate pull request, because the system is really driven through JSON and its native capabilities. What are your thoughts?

**Comment by znarf** [comment]
You're right, the upload needs to be refactored to handle non-JSON files, especially for the validation.

If you're good with it, no problem to merge the PR as it is. I'm just gonna have a last check and see if there is anything we can optimize.

**Comment by joncloud** [comment]
Go for it! Let me know if there is anything else I can do to help.

File: src/lib/github.js (hunk 1 above)
Comment (znarf) [reply src/lib/github.js]: We can, eg: `fetchWithOctokit('search.code', params, accessToken)`. Let's do it.

**Comment by znarf** [comment]
@joncloud as I'm testing the feature, looks like we're easily reaching the GitHub search API limit which is a concern.

GitHub is only allowing 30 requests per minute, https://developer.github.com/v3/search/ which is then immediately reached for an account with 15 repositories.

We need to think about that.

File: src/lib/dependencies/nuget.js [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    const cacheKey = `repo_nuget_dependencies_${githubRepo.id}`;
+    if (cache.has(cacheKey)) {
+      return cache.get(cacheKey);
+    }
+    
+    function mapPackages(searchPattern, transform) {
+      return searchFilesFromRepo(githubRepo, searchPattern, githubAccessToken)
+        .then(files => files.map(xml => new xmldoc.XmlDocument(xml))
+          .map(transform)
+          .reduce(aggregateDependencies)
```
Comment (znarf) [inline src/lib/dependencies/nuget.js]: This can lead to exceptions: `error: Reduce of empty array with no initial value`

**Comment by joncloud** [comment]
A couple thoughts around mitigating the API limit might be:
 * Requiring to sign in to enable NuGet
 * Queuing NuGet requests (synchronous or in the background)

Do you have any other thoughts on the subject?

**Comment by znarf** [comment]
Requiring to Sign In might help, because the limit is then 30 queries / minutes per authenticated user, but it's even low in that case. Also, it would be more complicated to explain: "Sign In to detect more dependencies (NuGet, etc ...). Why not?

Doing background requests might be the way to go, but we would need to make the architecture way more complex. We need to carefuly think if this is where we want to go.

Would be so great if we could rely on GitHub to extract dependencies with a single API call, using that:
https://help.github.com/articles/listing-the-packages-that-a-repository-depends-on/ And if this was Open Source so we can help them supporting more package managers.

**Comment by znarf** [comment]
I have put more though into this. I know think the proper short term solution is to do less requests to the GitHub search API.

- let's only fetch dependencies if the detected language is `C#`. I have already added that for the other languages, see: https://github.com/opencollective/backyourstack/commit/59807b1e8b4c8904fa1d31e2703617e22ffaf43c
- only search for `packages.config` if no `*.csproj` is detected.

With these changes, I think we can have something we can ship.

The only extra thing I would like to see added is a better handling of rate limiting errors. That's out of this PR scope and we can track that in https://github.com/opencollective/backyourstack/issues/61

File: src/lib/github.js (hunk 1 above)
Comment (joncloud) [reply src/lib/github.js]: Updated - this makes it much cleaner!

File: src/lib/dependencies/nuget.js [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+        .catch(err => {
+          _debug(`getDependenciesFromGithubRepo error: ${err.message}`);
+          return [];
+        });
+    }
+
+    return Promise.all(
+        mapPackages('*.csproj', csprojDependenciesStats),
+        mapPackages('packages.config', packagesConfigDependenciesStats)
+      )
```
[the remark this answers is not in the document]
Comment (joncloud) [reply src/lib/dependencies/nuget.js]: I've made a refactor to this piece that better aligns with the API call reduction, which ended up resolving this piece as well.

File: src/lib/dependencies/nuget.js (hunk 2 above)
Comment (joncloud) [reply src/lib/dependencies/nuget.js]: I've updated this to handle empty arrays.

**Comment by znarf** [comment]
Excellent. Merging!

**Comment by marcandre** [comment]
@joncloud Regarding support for file uploads, check #72

**Comment by znarf** [comment]
@joncloud It's now released and we made sure to detect the relevant Open Collectives, see https://github.com/opencollective/backyourstack/commit/9a02ce678f4d541fc9e2a301fac91a9ee5ff8acd

Thank you!

**Pull request merged**